actors/util/adt/map.go

```diff
@@ -27,48 +29,50 @@ type Map struct {
 	store Store
 }
 
-// NewMap creates a new HAMT with root `r` and store `s`.
-func NewMap(s Store, r cid.Cid) *Map {
+// AsMap interprets a store as a HAMT-based map with root `r`.
+func AsMap(s Store, r cid.Cid) *Map {
 	return &Map{
 		root:  r,
 		store: s,
 	}
 }
 
+// Creates a new map backed by an empty HAMT and flushes it to the store.
+func MakeEmptyMap(s Store) (*Map, error) {
+	nd := hamt.NewNode(s)
+	newMap := AsMap(s, cid.Undef)
+	err := newMap.write(nd)
+	return newMap, err
+}
+
 // Root return the root cid of HAMT.
 func (h *Map) Root() cid.Cid {
 	return h.root
 }
 
 // Put adds value `v` with key `k` to the hamt store.
 func (h *Map) Put(k Keyer, v vmr.CBORMarshaler) error {
-	oldRoot, err := hamt.LoadNode(h.store.Context(), h.store, h.root)
+	root, err := hamt.LoadNode(h.store.Context(), h.store, h.root)
 	if err != nil {
 		return errors.Wrapf(err, "Map Put failed to load node %v", h.root)
 	}
-	if err := oldRoot.Set(h.store.Context(), k.Key(), v); err != nil {
+	if err = root.Set(h.store.Context(), k.Key(), v); err != nil {
 		return errors.Wrapf(err, "Map Put failed set in node %v with key %v value %v", h.root, k.Key(), v)
 	}
-	if err := oldRoot.Flush(h.store.Context()); err != nil {
+	if err = root.Flush(h.store.Context()); err != nil {
 		return errors.Wrapf(err, "Map Put failed to flush node %v : %v", h.root, err)
 	}
 
-	// update the root
-	newRoot, err := h.store.Put(h.store.Context(), oldRoot)
-	if err != nil {
-		return errors.Wrapf(err, "Map Put failed to persist changes to store %s", h.root)
-	}
-	h.root = newRoot
-	return nil
+	return h.write(root)
 }
 
 // Get puts the value at `k` into `out`.
 func (h *Map) Get(k Keyer, out vmr.CBORUnmarshaler) (bool, error) {
-	oldRoot, err := hamt.LoadNode(h.store.Context(), h.store, h.root)
+	root, err := hamt.LoadNode(h.store.Context(), h.store, h.root)
 	if err != nil {
 		return false, errors.Wrapf(err, "Map Get failed to load node %v", h.root)
 	}
-	if err := oldRoot.Find(h.store.Context(), k.Key(), out); err != nil {
+	if err := root.Find(h.store.Context(), k.Key(), out); err != nil {
 		if err == hamt.ErrNotFound {
 			return false, nil
 		}
```

Comment:
Changes I made while rebasing, I think its useful to return a boolean here for the case that the key was not found in the map.